### PR TITLE
[#21] Add tests for models

### PR DESCRIPTION
**Pull request by Voovaaa**

### CODE REVIEW

File: first_django_project/discography/views.py [hunk 1]
```diff
@@ -27,11 +28,16 @@ class TrackCreateView(generic.CreateView):
     model = Track
     template_name = "discography/new_track.html"
     form_class = TrackForm
-    success_url = "/"
+
+    def form_valid(self, form):
+        track = form.save(commit=False)
+        track.artist_id = self.kwargs['pk']
+        track.save()
+        return redirect(reverse("discography:artist_detail", args=[track.artist_id]))
```
Comment (Safintim) [inline first_django_project/discography/views.py]: Use only redirect

File: first_django_project/discography/views.py [hunk 2]
```diff
@@ -27,11 +28,16 @@ class TrackCreateView(generic.CreateView):
     model = Track
     template_name = "discography/new_track.html"
     form_class = TrackForm
-    success_url = "/"
+
+    def form_valid(self, form):
+        track = form.save(commit=False)
+        track.artist_id = self.kwargs['pk']
```
Comment (Safintim) [inline first_django_project/discography/views.py]: rename pk to artist_pk

File: first_django_project/discography/models.py [hunk 3]
```diff
@@ -1,9 +1,16 @@
 from django.db import models
 from django.core.validators import MinLengthValidator
+from django.utils import timezone
+from django.core.exceptions import ValidationError
+
+
+def is_not_in_future_validator(value):
```
Comment (Safintim) [inline first_django_project/discography/models.py]: rename validate_date_published

File: first_django_project/discography/models.py [hunk 4]
```diff
@@ -1,9 +1,16 @@
...
+
+
+def is_not_in_future_validator(value):
+    if value > timezone.now().date():
+        raise ValidationError(f"Date published must be > {timezone.now().date()}")
 
 
 class Artist(models.Model):
-    name = models.CharField(max_length=100)
+    name = models.CharField(max_length=100, unique=True)
```
Comment (Safintim) [inline first_django_project/discography/models.py:13]: Where migrations?

File: first_django_project/discography/models.py [hunk 5]
```diff
@@ -12,18 +19,27 @@ def __str__(self):
...
+        unique_together = ('name', 'artist')
+        ordering = ['date_published']
+
 
 class Comment(models.Model):
     text = models.CharField(max_length=1000,
                             validators=[MinLengthValidator(3, message="Comment's len must be more than 2")])
     date_published = models.DateTimeField(auto_now_add=True)
     artist = models.ForeignKey(Artist, on_delete=models.CASCADE, related_name="comments")
+    parent_comment = models.ForeignKey("Comment", on_delete=models.CASCADE,
```
Comment (Safintim) [inline first_django_project/discography/models.py]: simple parent

File: first_django_project/discography/models.py [hunk 6]
```diff
@@ -12,18 +19,27 @@ def __str__(self):
...
+                                       related_name="child_comments", blank=True, null=True)
```
Comment (Safintim) [inline first_django_project/discography/models.py]: simple children

File: first_django_project/discography/tests/tests.py [hunk 7]
```diff
@@ -0,0 +1,68 @@
+from django.test import TestCase
+from django.urls import reverse
+from ..models import Artist, Track
```
Comment (Safintim) [inline first_django_project/discography/tests/tests.py:3]: use from _app_name_.models import

File: first_django_project/discography/tests/tests.py [hunk 8]
```diff
@@ -0,0 +1,68 @@
+from django.test import TestCase
+from django.urls import reverse
+from ..models import Artist, Track
+import datetime as dt
+
+
+class TestViewArtistDetail(TestCase):
+    def test_get(self):
+        artist = Artist.objects.create(name="artist1")
+        t1 = Track.objects.create(name="track1", date_published=dt.date(year=1,month=1,day=1), artist=artist, pk=1)
```
Comment (Safintim) [inline first_django_project/discography/tests/tests.py]: don't shorten variable name

Comment (Safintim) [reply first_django_project/discography/tests/tests.py]: why use pk=1?
use date_published="2020-02-01"

File: first_django_project/discography/tests/tests.py [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+from ..models import Artist, Track
+import datetime as dt
+
+
+class TestViewArtistDetail(TestCase):
+    def test_get(self):
+        artist = Artist.objects.create(name="artist1")
+        t1 = Track.objects.create(name="track1", date_published=dt.date(year=1,month=1,day=1), artist=artist, pk=1)
+        t2 = Track.objects.create(name="track2", date_published=dt.date(year=2,month=1,day=1), artist=artist, pk=2)
+        url = reverse("discography:artist_detail", args=[1])
```
Comment (Safintim) [inline first_django_project/discography/tests/tests.py]: what is "1"

File: first_django_project/discography/tests/tests.py [hunk 10]
```diff
@@ -0,0 +1,68 @@
...
+        self.assertEqual(expected_tracks, actual_tracks)
+
+
+class TestViewIndex(TestCase):
+    def test_get(self):
+        url = reverse("discography:index")
+
+        response = self.client.get(url)
+
+        expected_artists = list(Artist.objects.values_list("id", flat=True))
```
Comment (Safintim) [inline first_django_project/discography/tests/tests.py:33]: expected = []

File: first_django_project/discography/tests/tests.py [hunk 11]
```diff
@@ -0,0 +1,68 @@
...
+        url = reverse("discography:new_artist")
+
+        artist_name = "Artist New"
+
+        response = self.client.post(url, data={"name": artist_name})
+
+        expected_success_url = reverse("discography:index")
+        is_artist_exists = Artist.objects.filter(name=artist_name).exists()
+
+        # self.assertEqual(response.status_code, 200)
```
Comment (Safintim) [inline first_django_project/discography/tests/tests.py]: why comment

File: first_django_project/discography/tests/tests.py [hunk 12]
```diff
@@ -0,0 +1,68 @@
...
+class TestViewArtistCreate(TestCase):
+    def test_create(self):
+        url = reverse("discography:new_artist")
+
+        artist_name = "Artist New"
+
+        response = self.client.post(url, data={"name": artist_name})
+
+        expected_success_url = reverse("discography:index")
+        is_artist_exists = Artist.objects.filter(name=artist_name).exists()
```
Comment (Safintim) [inline first_django_project/discography/tests/tests.py]: is_created

File: first_django_project/discography/tests/tests.py [hunk 13]
```diff
@@ -0,0 +1,68 @@
+from django.test import TestCase
+from django.urls import reverse
+from ..models import Artist, Track
+import datetime as dt
+
+
+class TestViewArtistDetail(TestCase):
```
Comment (Safintim) [inline first_django_project/discography/tests/tests.py]: Test + View_name -> TestArtistDetailView

File: first_django_project/discography/tests/tests.py [hunk 14]
```diff
@@ -0,0 +1,68 @@
...
+        # self.assertEqual(response.status_code, 200)
+        self.assertRedirects(response, expected_success_url)
+        self.assertTrue(is_artist_exists)
+
+
+class TestViewTrackCreate(TestCase):
+    def test_create(self):
+        artist = Artist.objects.create(name="artist")
+        track_name = "track"
+        track_DP = dt.date(year=1,month=1,day=1)
```
Comment (Safintim) [inline first_django_project/discography/tests/tests.py]: don't shorten

File: first_django_project/discography/tests/tests.py [hunk 15]
```diff
@@ -0,0 +1,68 @@
...
+
+        url = reverse("discography:new_track", args=[1])
```
Comment (Safintim) [inline first_django_project/discography/tests/tests.py]: what is "1"

File: first_django_project/discography/tests/tests.py [hunk 16]
```diff
@@ -0,0 +1,68 @@
...
+
+        response = self.client.post(url, data={"name": track_name, "date_published": track_DP, "artist": artist})
```
Comment (Safintim) [inline first_django_project/discography/tests/tests.py]: create variable data = {}

File: first_django_project/discography/views.py (hunk 2 above)
Comment (Voovaaa) [reply first_django_project/discography/views.py]: and in urls.py

File: first_django_project/discography/tests/tests.py (hunk 9 above)
Comment (Voovaaa) [reply first_django_project/discography/tests/tests.py]: artist.pk

File: first_django_project/discography/tests/tests.py (hunk 11 above)
Comment (Voovaaa) [reply first_django_project/discography/tests/tests.py]: https://docs.python.org/3/library/http.html

File: first_django_project/discography/tests/tests.py (hunk 16 above)
Comment (Voovaaa) [reply first_django_project/discography/tests/tests.py]: create data as variable

File: first_django_project/discography/tests/tests.py [hunk 17]
```diff
@@ -0,0 +1,72 @@
...
+    def test_get(self):
+        artist = Artist.objects.create(name="artist1")
+        track1 = Track.objects.create(name="track1", date_published="0001-1-1", artist=artist)
+        track2 = Track.objects.create(name="track2", date_published="0002-2-2", artist=artist)
+        url = reverse("discography:artist_detail", args=[artist.pk])
+
+        response = self.client.get(url)
+
+        expected_tracks = set((Track.objects.filter(artist=artist)))
+        actual_tracks = {track1, track2}
```
Comment (Safintim) [inline first_django_project/discography/tests/tests.py:18]: why? why is not from context?

File: first_django_project/discography/tests/tests.py [hunk 18]
```diff
@@ -0,0 +1,72 @@
...
+
+        self.assertEqual(response.status_code, HTTPStatus.OK)
+        self.assertTemplateUsed(response, "discography/artist_detail.html")
+        self.assertEqual(expected_tracks, actual_tracks)
+
+
+class TestViewIndex(TestCase):
```
Comment (Safintim) [inline first_django_project/discography/tests/tests.py]: Name

File: first_django_project/discography/tests/tests.py [hunk 19]
```diff
@@ -0,0 +1,72 @@
...
+
+        expected_artists = list(Artist.objects.values_list("id", flat=True))
+        actual_artists = list(response.context_data["artists"].values_list("id", flat=True))
+
+        self.assertEqual(response.status_code, HTTPStatus.OK)
+        self.assertTemplateUsed(response, "discography/index.html")
+        self.assertEqual(actual_artists, expected_artists)
+
+
+class TestViewArtistCreate(TestCase):
```
Comment (Safintim) [inline first_django_project/discography/tests/tests.py]: Name

File: first_django_project/discography/tests/tests.py [hunk 20]
```diff
@@ -0,0 +1,72 @@
...
+
+        expected_success_url = reverse("discography:index")
+        is_created = Artist.objects.filter(name=artist_name).exists()
+
+        self.assertEqual(response.status_code, 200)
+        self.assertRedirects(response, expected_success_url)
+        self.assertTrue(is_created)
+
+
+class TestViewTrackCreate(TestCase):
```
Comment (Safintim) [inline first_django_project/discography/tests/tests.py]: Name

File: first_django_project/discography/tests/tests.py [hunk 21]
```diff
@@ -0,0 +1,72 @@
...
+        artist = Artist.objects.create(name="artist")
+        name = "track"
+        date_published = "0001-01-01"
+        data = {"name": name, "date_published": date_published, "artist": artist}
+
+        url = reverse("discography:new_track", args=[artist.pk])
+
+        response = self.client.post(url, data=data)
+
+        expected_succes_url = reverse("discography:artist_detail", args=[1])
```
Comment (Safintim) [inline first_django_project/discography/tests/tests.py]: args=[artist.pk]

File: first_django_project/discography/tests/tests.py [hunk 22]
```diff
@@ -0,0 +1,72 @@
...
+        self.assertRedirects(response, expected_success_url)
+        self.assertTrue(is_created)
+
+
+class TestViewTrackCreate(TestCase):
+    def test_create(self):
+        artist = Artist.objects.create(name="artist")
+        name = "track"
+        date_published = "0001-01-01"
+        data = {"name": name, "date_published": date_published, "artist": artist}
```
Comment (Safintim) [inline first_django_project/discography/tests/tests.py]: don't send artist

File: first_django_project/discography/tests/tests.py [hunk 23]
```diff
@@ -0,0 +1,72 @@
...
+        url = reverse("discography:new_artist")
+
+        artist_name = "Artist New"
+
+        response = self.client.post(url, data={"name": artist_name})
+
+        expected_success_url = reverse("discography:index")
+        is_created = Artist.objects.filter(name=artist_name).exists()
+
+        self.assertEqual(response.status_code, 200)
```
Comment (Safintim) [inline first_django_project/discography/tests/tests.py]: 200 - magic number

File: first_django_project/discography/tests/tests.py [hunk 24]
```diff
@@ -0,0 +1,72 @@
...
+        date_published = "0001-01-01"
+        data = {"name": name, "date_published": date_published, "artist": artist}
+
+        url = reverse("discography:new_track", args=[artist.pk])
+
+        response = self.client.post(url, data=data)
+
+        expected_succes_url = reverse("discography:artist_detail", args=[1])
+        track_exists = Track.objects.filter(name=name, date_published=date_published, artist=artist).exists()
+
```
Comment (Safintim) [inline first_django_project/discography/tests/tests.py:70]: check status code

File: first_django_project/discography/views.py [hunk 25]
```diff
@@ -18,7 +19,8 @@ class ArtistDetailView(generic.CreateView, generic.DetailView):
 
     def form_valid(self, form):
         comment = form.save(commit=False)
-        comment.artist_id = self.kwargs['pk']
+        artist_pk = self.kwargs['pk']
+        comment.artist_id = artist_pk
         comment.save()
         return redirect('discography:artist_detail', pk=self.kwargs['pk'])
```
Comment (Safintim) [inline first_django_project/discography/views.py:25]: pk=artist_pk

File: first_django_project/discography/views.py [hunk 26]
```diff
@@ -27,11 +29,17 @@ class TrackCreateView(generic.CreateView):
...
     template_name = "discography/new_track.html"
     form_class = TrackForm
-    success_url = "/"
+
+    def form_valid(self, form):
+        track = form.save(commit=False)
+        artist_pk = self.kwargs['pk']
+        track.artist_id = artist_pk
+        track.save()
+        return redirect(reverse("discography:artist_detail", args=[track.artist_id]))
```
Comment (Safintim) [inline first_django_project/discography/views.py]: Use only redirect

**Review by Safintim — APPROVED** [review]

**Pull request merged**